text/0003-templates.md

```diff
@@ -0,0 +1,447 @@
+- **Copying ambiguity**: Certain fields such as `version` may not be desirable to inherit from a parent `package.json`. The `extends` flavor would need to exclude copying for some fields. This makes a theoretical `extends` flavor difficult to understand and result in surprises at publish time.
+- **Merging ambiguity**: It's not always clear how to "_merge_" fields. For example, if a consuming package and template both define `pnpm.allowedDeprecatedVersions.express`, should pnpm replace the field from the template with the child package's value, or union the allowed deprecated versions? What should `extends` do for fields added in newer versions of pnpm that it doesn't know how to merge?
+
+A future version of pnpm templates may provide this flavor, but significant thoughtfulness around the above would be necessary. For now, we believe all use cases involving an `extends` flavor could be addressed through the introduction of new template flavors.
+
+## Implementation
+
+### Fetching
+
+Most templates are not installed as standard dependencies and linked into `node_modules`. If a package refers to an external template on an NPM registry, only the metadata will be fetched. The fetch will be performed [without the abbreviated header](https://github.com/npm/registry/blob/master/docs/responses/package-metadata.md#abbreviated-metadata-format), so the full document is retrieved. The metadata will be saved to the content-addressable store for fast subsequent lookups.
+
+For the `catalog` template, not installing the template into `node_modules` provides a performance optimization. Only the subset of the version catalog that's used will be fetched and installed.
+
+The exception is the `patches` template flavor since it may refer to local patch files. The full NPM package producing the `patches` template will be fetched, added to the pnpm store, and linked into `node_modules/.pnpm`.
+
+### Lockfile
+
+The relevant sections of each template will be saved to `pnpm-lock.yaml` under a new `templates` key. This allows users to more easily review changes to templates and pnpm to perform faster up-to-date checks.
+
+```yaml
+lockfileVersion: next
+
+importers:
+  # ...
+
+templates:
+
+  /@example/frontend-catalog@0.1.0:
+    manifestResolution: {integrity: sha512...}
+    pnpm:
+      template: catalog
+    dependencies:
+      react: ^18.2.0
+      react-dom: ^18.2.0
+      redux: ^4.2.0
+      react-redux: ^8.0.0
+
+  /@example/patches@0.1.0:
+    resolution: {integrity: sha512...}
+    pnpm:
+      template: patches
+    packageExtensions:
+      react-redux:
+        peerDependencies:
+          react-dom: "*"
+    patchedDependencies:
+      express@4.18.1: patches/express@4.18.1.patch
+
+packages:
+  # ...
+```
+
+Since an integrity checksum of only the manifest is not provided by the registry, a new `manifestResolution.integrity` value will need to be computed locally.
```

Comment:
Saving only the integrity of the manifest is needed since we're not fetching the full tarball. This might be a bad idea for reasons I haven't thought of yet. Perhaps we should simply fetch the full tarball and read the `package.json` inside, despite the registry providing the ability to fetch only the manifest.

Reply:
are you sure we won't download the tarball? What if I want to share a patch file with the template package?

Reply:
That's right — we do have to download the full tarball for `patches` templates. This is currently described as an exception specifically for patches:

https://github.com/pnpm/rfcs/blob/cb35007544bce664af0387ff93d93bcb7a194275/text/0003-templates.md?plain=1#L376

For patches we use the standard `resolution` block rather than `manifestResolution`.

https://github.com/pnpm/rfcs/blob/cb35007544bce664af0387ff93d93bcb7a194275/text/0003-templates.md?plain=1#L400-L401

But to be honest I'm not too sure of this approach. We could eliminate complexity by always fetching the tarball regardless of what template flavor/type is being requested.

Reply:
even if we don't download the tarball, we still can use the tarball's integrity checksum as it is stored in the metadata. There's no need to download the tarball to get its integrity.

Reply:
Updated in commit https://github.com/pnpm/rfcs/pull/3/commits/22da4acd3409e7d0075903ea11238d92409d05f9.

Had to think through this a bit, but I think the idea of storing the tarball's integrity is solid (even if it's unused). 👍 Thanks for the suggestion. I was worried about malicious edits to the lockfile or malicious registries. But malicious users would just would just modify the integrity hash anyway when editing fields, and all bets are off in the case of a malicious registry without E2E signing.